src/main/java/bio/terra/janitor/service/cleanup/flight/GoogleUtils.java

```diff
@@ -2,15 +2,95 @@
 public class GoogleUtils {
   private GoogleUtils() {}
 
+  private static final String GET_PROJECT_PERMISSION = "resourcemanager.projects.get";
+
+  /**
+   * Try to get a project that may not exist.
+   *
+   * <p>We cannot distinguish between not having access to a project and the project no longer
+   * existing. Google returns 403 for both cases to prevent project id probing.
+   *
+   * <p>For now, we think that due to how long it takes to delete a project vs marking it as ready
+   * for deletion, if we know that we have permissions on the parent resource containing the
+   * project, the project probably never existed. If we do not know the parent resource, or do not
+   * have permissions on the project we assume a 403 is truly foribdden.
+   */
+  public static ProjectStatus checkProjectStatus(
+      String projectId, Optional<String> parentResource, CloudResourceManagerCow resourceManager)
+      throws IOException {
+    try {
+      Project project = resourceManager.projects().get(projectId).execute();
+      // We were able to retrieve the project.
+      return deleteInProgress(project) ? ProjectStatus.DELETE_IN_PROGRESS : ProjectStatus.ACTIVE;
+    } catch (GoogleJsonResponseException e) {
+      if (e.getStatusCode() != HttpStatus.SC_FORBIDDEN) {
+        // Got an unexpected exception, throw to the caller.
+        throw e;
+      }
+    }
+    // If we reach here, we were forbidden from retrieving the project. We now try to determine if
+    // we have enough permissions to assume the project does not exist.
+    if (!parentResource.isEmpty()
+        && hasProjectGetPermission(parentResource.get(), resourceManager)) {
+      return ProjectStatus.PROBABLY_DOES_NOT_EXIST;
+    } else {
+      return ProjectStatus.FORBIDDEN;
+    }
+  }
+
+  /**
+   * Returns whether the credentials on the {@link CloudResourceManagerCow} have permissions to get
+   * projects on the folder.
+   */
+  private static boolean hasProjectGetPermission(
+      String parentResource, CloudResourceManagerCow resourceManager) throws IOException {
+    if (!parentResource.contains("folders/")) {
+      // We only support checking for permissions on folders right now.
+      return false;
+    }
+    TestIamPermissionsResponse iamResponse =
+        resourceManager
+            .folders()
+            .testIamPermissions(
+                parentResource,
+                new TestIamPermissionsRequest()
+                    .setPermissions(ImmutableList.of(GET_PROJECT_PERMISSION)))
+            .execute();
+    return iamResponse.getPermissions().contains(GET_PROJECT_PERMISSION);
+  }
+
+  /**
+   * Our best determination of a project that may be deleted status. See {@link
+   * #checkProjectStatus(String, Optional, CloudResourceManagerCow)}.
+   */
+  enum ProjectStatus {
+    // The project is exists and is active.
+    ACTIVE,
+    // The project is in the process of being deleted. It is partially unavailable.
+    DELETE_IN_PROGRESS,
```

Comment:
GCP Lifecycle also has this name. And we use the same name but mapped to two GCP status:
`DELETE_REQUESTED` and `DELETE_IN_PROGRESS`. This confuses me if I did not trace whether its Janitor or GCP's  `DELETE_IN_PROGRESS`. Using different name to summarize those two GCP status might be helpful?  Like `DELETING`?

Reply:
Good idea, done.